### PR TITLE
[ml/train] `TorchTrainer` implementation

### PR DESCRIPTION
**Pull request by amogkam**
Depends on #23211 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/ml/BUILD [hunk 1]
```diff
@@ -1,3 +1,57 @@
...
+# Tests from the python/ray/ml/examples directory.
+# Please keep these sorted alphabetically.
+# --------------------------------------------------------------------
+py_test(
+    name = "custom_trainer",
+    size = "small",
+    main = "examples/custom_trainer.py",
+    srcs = ["examples/custom_trainer.py"],
+    tags = ["team:ml", "exclusive"],
+    deps = [":train_lib"]
```
Comment (Yard1) [inline python/ray/ml/BUILD:11]: `:ml_lib`?

**Review by richardliaw — APPROVED** [review]
Pretty big fan of the new API. Reviewed just the torch code. Looks good to me, though the last reviewer should actually review the rebased PR.

File: python/ray/ml/BUILD (hunk 1 above)
Comment (amogkam) [reply python/ray/ml/BUILD:11]: Yes you are 100% correct, it should be ml_lib!

Comment (amogkam) [reply python/ray/ml/BUILD:11]: Updated

**Review by ericl — APPROVED** [review]

**Review by matthewdeng — APPROVED** [review]
Nice!

**Pull request merged**